plugins/inputs/snmp_trap/snmp_trap.go

```diff
@@ -0,0 +1,144 @@
+package snmp_trap
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/snmp"
+
+	"github.com/soniah/gosnmp"
+)
+
+type SnmpTrap struct {
+	Port uint16 `toml:"port"`
+
+	acc      telegraf.Accumulator
+	listener *gosnmp.TrapListener
+	wg       sync.WaitGroup
+	timeFunc func() time.Time
+
+	makeHandlerWrapper func(func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr)) func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr)
+	Errch              chan error
+}
+
+var sampleConfig = `
+  ## Port to listen on.  Default 162
+  #port = 162
+`
+
+func (s *SnmpTrap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *SnmpTrap) Description() string {
+	return "Receive SNMP traps"
+}
+
+func (s *SnmpTrap) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func init() {
+	inputs.Add("snmp_trap", func() telegraf.Input {
+		return &SnmpTrap{
+			timeFunc: time.Now,
+			Port:     162,
+		}
+	})
+}
+
+func (s *SnmpTrap) Init() error {
+	return nil
+}
+
+func (s *SnmpTrap) Start(acc telegraf.Accumulator) error {
+	s.acc = acc
+	s.listener = gosnmp.NewTrapListener()
+	s.listener.OnNewTrap = makeTrapHandler(s)
+	s.listener.Params = gosnmp.Default
+	//s.listener.Params.Logger = log.New(os.Stdout, "", 0)
+
+	//wrap the handler, used in unit tests
+	if nil != s.makeHandlerWrapper {
+		s.listener.OnNewTrap = s.makeHandlerWrapper(s.listener.OnNewTrap)
+	}
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		//no ip means listen on all interfaces, ipv4 and ipv6
+		err := s.listener.Listen(":" + strconv.FormatUint(uint64(s.Port), 10))
+		if err != nil {
+			s.Errch <- err
+			log.Panicf("error in listen: %s", err)
+		}
+	}()
+
+	return nil
+}
+
+func (s *SnmpTrap) Stop() {
+	s.listener.Close()
+	s.wg.Wait()
+}
+
+func makeTrapHandler(s *SnmpTrap) func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr) {
+	return func(packet *gosnmp.SnmpPacket, addr *net.UDPAddr) {
+		tm := s.timeFunc()
+		fields := map[string]interface{}{}
+		tags := map[string]string{}
+
+		tags["trap_version"] = packet.Version.String()
+
+		for _, v := range packet.Variables {
+			//build a name and value for each variable to use as tags
+			//and fields.  defaults are the uninterpreted values
+			name := v.Name
```

Comment:
I'm not sure what is returned as a name here, but would it be preferred to snake case them to fit better as field/tag keys?

Reply:
v.Name is a numeric OID, something like ".1.3.6.1.6.3.1.1.5.1".  If MIB lookup is set up we'll get something like "coldStart" that we use instead.  SNMP convention is camel case and well defined in many MIBs over many decades.  I was going to leave field names as camel case.  If there's enough demand we could consider a configuration option to automatically convert to snake case.

Reply:
Normally we use snake_case in Telegraf, even when the source system uses another style, but for this plugin I think we should mirror the snmp input behavior.